File: go/admin_test.go
package realmid

import (
	"context"
	"encoding/json"
	"net/http"
	"net/http/httptest"
	"net/url"
	"testing"
)

func adminPlatformTokenStub(mux *http.ServeMux) {
	mux.HandleFunc("/auth/login", func(w http.ResponseWriter, _ *http.Request) {
		_ = json.NewEncoder(w).Encode(map[string]any{"status": "ok", "subject_type": "platform", "refresh_token": "rtok-platform", "access_token": "ptok", "expires_in": 300})
	})
}

func TestAdmin_ListPlatforms_QueryAndDecode(t *testing.T) {
	mux := http.NewServeMux()
	adminPlatformTokenStub(mux)
	var gotQ url.Values
	mux.HandleFunc("/admin/platforms", func(w http.ResponseWriter, r *http.Request) {
		gotQ = r.URL.Query()
		next := "n2"
		_ = json.NewEncoder(w).Encode(map[string]any{
			"items": []any{
				map[string]any{
					"id":           "p1",
					"display_name": "Acme",
					"slug":         "acme",
					"status":       "active",
					"signup_mode":  "closed",
					"domains":      []string{"acme.test"},
					"owner": map[string]any{
						"user_id": "u1", "name": "Alice", "email": "a@acme.test",
					},
					"tenants_count":    3,
					"users_count":      9,
					"last_activity_at": 1700000000,
					"created_at":       1690000000,
				},
			},
			"next_cursor": next,
			"total":       42,
		})
	})
	srv := httptest.NewServer(mux)
	defer srv.Close()

	r, _ := NewRealm(Config{RealmID: testRealmID, APIKey: "rk", BaseURL: srv.URL})
	hcd := true
	out, err := r.Admin.ListPlatforms(context.Background(), ListPlatformsParams{
		Q:               "ac",
		Status:          []string{"active", "suspended"},
		SignupMode:      []string{"closed"},
		Domain:          "acme.test",
		HasCustomDomain: &hcd,
		CreatedAfter:    1680000000,
		Sort:            "created_desc",
		Limit:           25,
	})
	if err != nil {
		t.Fatalf("ListPlatforms: %v", err)
	}
	if len(out.Items) != 1 || out.Items[0].ID != "p1" || out.Items[0].Owner.Email != "a@acme.test" {
		t.Errorf("unexpected items: %+v", out.Items)
	}
	if out.NextCursor == nil || *out.NextCursor != "n2" {
		t.Errorf("unexpected next_cursor: %+v", out.NextCursor)
	}
	if out.Total != 42 {
		t.Errorf("unexpected total: %d", out.Total)
	}
	if gotQ.Get("q") != "ac" || gotQ.Get("status") != "active,suspended" || gotQ.Get("has_custom_domain") != "true" {
		t.Errorf("unexpected query: %+v", gotQ)
	}
	if gotQ.Get("limit") != "25" || gotQ.Get("sort") != "created_desc" {
		t.Errorf("unexpected sort/limit: %+v", gotQ)
	}
}

func TestAdmin_Stats(t *testing.T) {
	mux := http.NewServeMux()
	adminPlatformTokenStub(mux)
	mux.HandleFunc("/admin/stats", func(w http.ResponseWriter, _ *http.Request) {
		_ = json.NewEncoder(w).Encode(map[string]any{
			"platforms_count": 4, "tenants_count": 12, "users_count": 88,
			"sessions_active": 7, "events_24h": 3,
		})
	})
	srv := httptest.NewServer(mux)
	defer srv.Close()

	r, _ := NewRealm(Config{RealmID: testRealmID, APIKey: "rk", BaseURL: srv.URL})
	s, err := r.Admin.Stats(context.Background())
	if err != nil {
		t.Fatalf("Stats: %v", err)
	}
	if s.PlatformsCount != 4 || s.UsersCount != 88 || s.Events24h != 3 {
		t.Errorf("unexpected stats: %+v", s)
	}
}

func TestAdmin_ListEvents(t *testing.T) {
	mux := http.NewServeMux()
	adminPlatformTokenStub(mux)
	var gotQ url.Values
	mux.HandleFunc("/admin/events", func(w http.ResponseWriter, r *http.Request) {
		gotQ = r.URL.Query()
		_ = json.NewEncoder(w).Encode(map[string]any{
			"items": []any{
				map[string]any{
					"id": 1, "occurred_at": 1700000000, "kind": "platform.created",
					"actor_user_id": "u1", "actor_label": "alice@acme",
					"platform_id": "p1", "summary": "alice@acme platform.created p1",
				},
			},
			"next_cursor": nil,
		})
	})
	srv := httptest.NewServer(mux)
	defer srv.Close()

	r, _ := NewRealm(Config{RealmID: testRealmID, APIKey: "rk", BaseURL: srv.URL})
	out, err := r.Admin.ListEvents(context.Background(), ListEventsParams{
		PlatformID: "p1",
		Kind:       []string{"platform.created", "platform.suspended"},
		Since:      1690000000,
		Limit:      50,
	})
	if err != nil {
		t.Fatalf("ListEvents: %v", err)
	}
	if len(out.Items) != 1 || out.Items[0].ID != 1 || out.Items[0].Kind != "platform.created" {
		t.Errorf("unexpected events: %+v", out.Items)
	}
	if out.NextCursor != nil {
		t.Errorf("expected nil next_cursor, got %v", *out.NextCursor)
	}
	if gotQ.Get("platform_id") != "p1" || gotQ.Get("kind") != "platform.created,platform.suspended" {
		t.Errorf("unexpected query: %+v", gotQ)
	}
}

func TestAdmin_Search(t *testing.T) {
	mux := http.NewServeMux()
	adminPlatformTokenStub(mux)
	var gotQ url.Values
	mux.HandleFunc("/admin/search", func(w http.ResponseWriter, r *http.Request) {
		gotQ = r.URL.Query()
		_ = json.NewEncoder(w).Encode(map[string]any{
			"items": []any{
				map[string]any{"type": "platform", "id": "p1", "label": "Acme", "sublabel": "acme"},
				map[string]any{"type": "tenant", "id": "t1", "label": "Acme Default", "platform_id": "p1"},
			},
		})
	})
	srv := httptest.NewServer(mux)
	defer srv.Close()

	r, _ := NewRealm(Config{RealmID: testRealmID, APIKey: "rk", BaseURL: srv.URL})
	out, err := r.Admin.Search(context.Background(), "ac", 10)
	if err != nil {
		t.Fatalf("Search: %v", err)
	}
	if len(out.Items) != 2 || out.Items[0].Type != "platform" || out.Items[1].PlatformID != "p1" {
		t.Errorf("unexpected hits: %+v", out.Items)
	}
	if gotQ.Get("q") != "ac" || gotQ.Get("limit") != "10" {
		t.Errorf("unexpected query: %+v", gotQ)
	}
}

func TestAdmin_ForwardsForbidden(t *testing.T) {
	mux := http.NewServeMux()
	adminPlatformTokenStub(mux)
	mux.HandleFunc("/admin/stats", func(w http.ResponseWriter, _ *http.Request) {
		w.WriteHeader(http.StatusForbidden)
		_ = json.NewEncoder(w).Encode(map[string]any{
			"error": map[string]any{"code": "forbidden", "message": "base-realm staff required"},
		})
	})
	srv := httptest.NewServer(mux)
	defer srv.Close()

	r, _ := NewRealm(Config{RealmID: testRealmID, APIKey: "rk", BaseURL: srv.URL})
	_, err := r.Admin.Stats(context.Background())
	if !IsCode(err, ErrCodeForbidden) {
		t.Errorf("expected forbidden, got %v", err)
	}
}

File: ts/src/admin.test.ts
import { test } from "node:test";
import { strict as assert } from "node:assert";
import { createRealm } from "./realm.js";
import { RealmError } from "./errors.js";

function platformTokenStub(input: RequestInfo | URL): Response | null {
  const url = typeof input === "string" ? input : input.toString();
  if (url.endsWith("/auth/login")) {
    return new Response(JSON.stringify({ status: "ok", subject_type: "platform", refresh_token: "rtok-platform", access_token: "pt_x", expires_in: 300}), {
      status: 200, headers: { "content-type": "application/json" },
    });
  }
  return null;
}

test("admin.listPlatforms: forwards filters and decodes envelope", async () => {
  let hitUrl = "";
  const fetch: typeof fetch = (async (input: RequestInfo | URL) => {
    const stub = platformTokenStub(input);
    if (stub) return stub;
    hitUrl = typeof input === "string" ? input : input.toString();
    return new Response(JSON.stringify({
      items: [{
        id: "p1", display_name: "Acme", slug: "acme",
        status: "active", signup_mode: "closed",
        domains: ["acme.test"],
        owner: { user_id: "u1", name: "Alice", email: "a@acme.test" },
        tenants_count: 3, users_count: 9,
        last_activity_at: 1700000000, created_at: 1690000000,
      }],
      next_cursor: "n2",
      total: 42,
    }), { status: 200, headers: { "content-type": "application/json" } });
  }) as typeof fetch;

  const realm = createRealm({ realmId: "r", apiKey: "rk_live_x", baseUrl: "https://auth.test", fetch });
  const out = await realm.admin.listPlatforms({
    q: "ac",
    status: ["active", "suspended"],
    signupMode: ["closed"],
    hasCustomDomain: true,
    createdAfter: 1680000000,
    sort: "created_desc",
    limit: 25,
  });
  assert.equal(out.items.length, 1);
  assert.equal(out.items[0]!.id, "p1");
  assert.equal(out.items[0]!.owner.email, "a@acme.test");
  assert.equal(out.next_cursor, "n2");
  assert.equal(out.total, 42);
  assert.match(hitUrl, /\/admin\/platforms\?/);
  assert.match(hitUrl, /q=ac/);
  assert.match(hitUrl, /status=active%2Csuspended/);
  assert.match(hitUrl, /signup_mode=closed/);
  assert.match(hitUrl, /has_custom_domain=true/);
  assert.match(hitUrl, /created_after=1680000000/);
  assert.match(hitUrl, /sort=created_desc/);
  assert.match(hitUrl, /limit=25/);
});

test("admin.stats: GETs /admin/stats and decodes", async () => {
  const fetch: typeof fetch = (async (input: RequestInfo | URL) => {
    const stub = platformTokenStub(input);
    if (stub) return stub;
    return new Response(JSON.stringify({
      platforms_count: 4, tenants_count: 12, users_count: 88,
      sessions_active: 7, events_24h: 3,
    }), { status: 200, headers: { "content-type": "application/json" } });
  }) as typeof fetch;
  const realm = createRealm({ realmId: "r", apiKey: "rk_live_x", baseUrl: "https://auth.test", fetch });
  const s = await realm.admin.stats();
  assert.equal(s.platforms_count, 4);
  assert.equal(s.events_24h, 3);
});

test("admin.listEvents: forwards kind+platform_id filters", async () => {
  let hitUrl = "";
  const fetch: typeof fetch = (async (input: RequestInfo | URL) => {
    const stub = platformTokenStub(input);
    if (stub) return stub;
    hitUrl = typeof input === "string" ? input : input.toString();
    return new Response(JSON.stringify({
      items: [{
        id: 1, occurred_at: 1700000000, kind: "platform.created",
        actor_user_id: "u1", actor_label: "alice@acme",
        platform_id: "p1", summary: "alice@acme platform.created p1",
      }],
      next_cursor: null,
    }), { status: 200, headers: { "content-type": "application/json" } });
  }) as typeof fetch;
  const realm = createRealm({ realmId: "r", apiKey: "rk_live_x", baseUrl: "https://auth.test", fetch });
  const out = await realm.admin.listEvents({
    platformId: "p1",
    kind: ["platform.created", "platform.suspended"],
    since: 1690000000,
    limit: 50,
  });
  assert.equal(out.items.length, 1);
  assert.equal(out.items[0]!.id, 1);
  assert.equal(out.next_cursor, null);
  assert.match(hitUrl, /platform_id=p1/);
  assert.match(hitUrl, /kind=platform.created%2Cplatform.suspended/);
  assert.match(hitUrl, /since=1690000000/);
});

test("admin.search: q + limit", async () => {
  let hitUrl = "";
  const fetch: typeof fetch = (async (input: RequestInfo | URL) => {
    const stub = platformTokenStub(input);
    if (stub) return stub;
    hitUrl = typeof input === "string" ? input : input.toString();
    return new Response(JSON.stringify({
      items: [
        { type: "platform", id: "p1", label: "Acme", sublabel: "acme" },
        { type: "tenant", id: "t1", label: "Acme Default", platform_id: "p1" },
      ],
    }), { status: 200, headers: { "content-type": "application/json" } });
  }) as typeof fetch;
  const realm = createRealm({ realmId: "r", apiKey: "rk_live_x", baseUrl: "https://auth.test", fetch });
  const out = await realm.admin.search("ac", 10);
  assert.equal(out.items.length, 2);
  assert.equal(out.items[0]!.type, "platform");
  assert.equal(out.items[1]!.platform_id, "p1");
  assert.match(hitUrl, /q=ac/);
  assert.match(hitUrl, /limit=10/);
});

test("admin.stats: surfaces 403 forbidden envelope as RealmError(forbidden)", async () => {
  const fetch: typeof fetch = (async (input: RequestInfo | URL) => {
    const stub = platformTokenStub(input);
    if (stub) return stub;
    return new Response(JSON.stringify({
      error: { code: "forbidden", message: "base-realm staff required" },
    }), { status: 403, headers: { "content-type": "application/json" } });
  }) as typeof fetch;
  const realm = createRealm({ realmId: "r", apiKey: "rk_live_x", baseUrl: "https://auth.test", fetch });
  await assert.rejects(() => realm.admin.stats(), (e: Error) => {
    return e instanceof RealmError && e.code === "forbidden";
  });
});

test("admin.getPlatform: GETs /admin/platforms/{id} and decodes the fleet row", async () => {
  let hitUrl = "";
  let hitMethod = "";
  let calls = 0;
  const fetch: typeof fetch = (async (input: RequestInfo | URL, init?: RequestInit) => {
    const stub = platformTokenStub(input);
    if (stub) return stub;
    calls++;
    hitUrl = typeof input === "string" ? input : input.toString();
    hitMethod = init?.method ?? "GET";
    return new Response(JSON.stringify({
      id: "p1", display_name: "Acme", slug: "acme",
      status: "active", signup_mode: "closed",
      domains: ["acme.test"],
      owner: { user_id: "u1", name: "Alice", email: "a@acme.test" },
      tenants_count: 3, users_count: 9,
      last_activity_at: 1700000000, created_at: 1690000000,
    }), { status: 200, headers: { "content-type": "application/json" } });
  }) as typeof fetch;

  const realm = createRealm({ realmId: "r", apiKey: "rk_live_x", baseUrl: "https://auth.test", fetch });
  const out = await realm.admin.getPlatform("p1");

  // The row is the SAME AdminPlatformSummary the list returns — single-sourced
  // server-side through one store query and one serializer, so the detail
  // screen and the fleet table cannot disagree about a platform.
  assert.equal(out.id, "p1");
  assert.equal(out.display_name, "Acme");
  assert.equal(out.owner.email, "a@acme.test");
  assert.equal(out.tenants_count, 3);
  assert.equal(out.last_activity_at, 1700000000);

  assert.equal(hitMethod, "GET");
  assert.match(hitUrl, /\/admin\/platforms\/p1$/);
  // No query string: this is the by-id read, not a filtered list.
  assert.ok(!hitUrl.includes("?"), `expected no query string, got ${hitUrl}`);
  // ONE call. The screen this backs previously paged the list up to 20 times;
  // a wrapper that fans out would reintroduce exactly that.
  assert.equal(calls, 1);
});

test("admin.getPlatform: percent-encodes the id into the path", async () => {
  let hitUrl = "";
  const fetch: typeof fetch = (async (input: RequestInfo | URL) => {
    const stub = platformTokenStub(input);
    if (stub) return stub;
    hitUrl = typeof input === "string" ? input : input.toString();
    return new Response("{}", { status: 200, headers: { "content-type": "application/json" } });
  }) as typeof fetch;
  const realm = createRealm({ realmId: "r", apiKey: "rk_live_x", baseUrl: "https://auth.test", fetch });
  await realm.admin.getPlatform("a/../b");
  // An unencoded id would escape the /admin/platforms/ prefix and address a
  // different endpoint entirely.
  assert.ok(!hitUrl.includes("a/../b"), `id leaked into the path unencoded: ${hitUrl}`);
  assert.match(hitUrl, /\/admin\/platforms\/a%2F..%2Fb$/);
});

test("admin.getPlatform: a 404 stays platform_not_found, never a forbidden flavour", async () => {
  const fetch: typeof fetch = (async (input: RequestInfo | URL) => {
    const stub = platformTokenStub(input);
    if (stub) return stub;
    return new Response(JSON.stringify({
      error: { code: "platform_not_found", message: "platform not found" },
    }), { status: 404, headers: { "content-type": "application/json" } });
  }) as typeof fetch;
  const realm = createRealm({ realmId: "r", apiKey: "rk_live_x", baseUrl: "https://auth.test", fetch });
  // The issuer returns an IDENTICAL 404 for an id that was never issued and for
  // a platform the caller may not see (issuer DECISIONS.md 2026-08-06). That
  // indistinguishability is the security property: a distinct refusal would
  // confirm the id is live. The wrapper must not re-label it as a permission
  // error, which is what a caller would reach for to render "you don't have
  // access to this platform" — and that string IS the oracle.
  await assert.rejects(() => realm.admin.getPlatform("nope"), (e: Error) => {
    if (!(e instanceof RealmError)) return false;
    // `platform_not_found` is REGISTERED as of ts 0.38.0 / go 0.46.0 /
    // java 0.36.0, so it now survives to `error.code` instead of falling back
    // to statusToCode(404). This assertion moved with it — it is the visible
    // half of that behaviour change, and the reason the release is flagged
    // BREAKING despite being purely additive to the union.
    assert.equal(e.code, "platform_not_found");
    assert.equal(e.httpStatus, 404);
    assert.notEqual(e.code, "forbidden");
    assert.notEqual(e.code, "unauthorized");
    return true;
  });
});
